doc/dotNET SDK Proposal.md

```diff
@@ -0,0 +1,232 @@
+
+Users provide workflow implementations and activity implementations in one of two ways:
+ - Pattern based: Entry points of workflow and activity implementations follow specific, well known patterns. Developers use attributes to provide metadata about such entry point to the host. The host automatically loads, and executes workflows and activities. This approach allows writing workflows (and activities) with hardly any scaffolding. 
+- Interface based: Developers implement specific interfaces to create workflows (and activities), and point the host to these implementations. The host automatically loads, and executes workflows and activities. This approach requires a little more scaffolding. However, it also allows additional functionality and helps addressing some advanced scenarios (e.g. dynamic workflow APIs).
+
+### Basic Workflow Execution
+
+Consider a sample workflow of type `SomeWorkflow` that runs a loop executing `SomeActivity` until a certain condition is met. When such condition is eventually met, the workflow is notified via a `NotifyConditionMet` signal.
+
+At application startup, the application is configured to host a worker that can execute workflows (and activities). Then, the worker is configured to poll `SomeTaskQueue` for work items, and to support execution of `SomeWorkflow` instances:
+
+```cs
+public static void Main(string[] args)
+{     
+    IHost appHost = Host.CreateDefaultBuilder(args)
+            .UseTemporalWorkerHost()
+            .ConfigureServices(serviceCollection =>
+            {
+                serviceCollection.AddTemporalWorker()
+                        .Configure(temporalWorkerConfig =>
+                        {
+                            temporalWorkerConfig.TaskQueueMoniker = "SomeTaskQueue";
+                        });
+
+                serviceCollection.AddWorkflowWithAttributes<SomeWorkflow>();
+            })
+            .Build();
+
+    appHost.Run();
+}
+```
+
+We use the `WorkflowAttribute` to mark the class that implements the workflow and to point to the method that implements the main workflow logic. The `WorkflowSignalHandler` is used to specify the method that handles `NotifyConditionMet`-signals:
+
+```cs
+[Workflow(runMethod: nameof(MainAsync))]
+public class SomeWorkflow
+{
+    private bool _isConditionMet = false;
+
+    public async Task MainAsync(WorkflowContext workflowCtx)
+    {
+        while (! _isConditionMet)
+        {
+            await workflowCtx.Orchestrator.Activities.ExecuteAsync("SomeActivity");
+        }                                
+    }
+
+    [WorkflowSignalHandler]
+    public void NotifyConditionMetAsync()
```

Comment:
It's convention and MSFT's guidance: https://docs.microsoft.com/en-us/dotnet/csharp/async
There are, however, flame wars and discussions on that:
e.g., 
https://github.com/dotnet/core/issues/1464
https://github.com/dotnet/roslyn/issues/47592
https://github.com/dotnet/runtime/issues/26908
and there is more.

I go with using it because it makes things some nicely explicit, and because we do not use async everywhere.
E.g. query handlers are not async.

Reply:
My vote is to remove it for our use case. In all other SDKs we allow the names of methods to mean something for the contract (e.g. signal handler), and suggesting a naming convention will force everyone to either let that suggestion invade the contract, or explicitly override the name of the thing on every definition.

For example, what is the signal name as it appears in history, UI, etc of:

```
[WorkflowSignalHandler]
public Task SetAddresseeAsync(SpeechRequest input, WorkflowContext _)
```

This is the same type of dogma suggested by the hungarian notation proponents. Good thing .Net didn't suggest we add the word "Int" to all integer variables.

Reply:
You can name the method whatever you want, as long as it has the right signature and the attribute. However, in the examples I follow the applicable convention. I'll add a section to clarify how the _signal type name_ is generated from the method name.